File: src/test/java/me/breidenbach/rabbitex/connection/RabbitConsumerTest.java
package me.breidenbach.rabbitex.connection;

import com.rabbitmq.client.*;
import me.breidenbach.rabbitex.MessageHandler;
import mockit.Deencapsulation;
import mockit.Expectations;
import mockit.Mocked;
import org.junit.Test;

import java.io.IOException;



import static junit.framework.Assert.assertNotNull;

/**
 * User: Kevin E. Breidenbach
 * Date: 11/3/13
 * Time: 10:44 AM
 * Copyright 2013 Kevin E. Breidenbach
 */
 public class RabbitConsumerTest {

    private static final String QUEUE = "queue";
    private static final String EXCHANGE = "exchange";
    private static final String SUBJECT = "subject";
    private static final String ERROR_EXCHANGE = "ERROR_EXCHANGE";
    private static final String ERROR_SUBJECT = "ERROR_SUBJECT";

    @Mocked RabbitConnection mockRabbitConnection;
    @Mocked Connection mockConnection;
    @Mocked Channel mockChannel;
    @Mocked MessageHandler mockHandler;

    @Test
    public void constructor() throws IOException, RabbitConnectionException {
        new Expectations() {
            {
                mockRabbitConnection.connection(); result = mockConnection;
                mockConnection.createChannel(); result = mockChannel;
                mockChannel.queueDeclare(QUEUE, true, false, false, null); times = 1;
                mockChannel.queueBind(QUEUE, EXCHANGE, SUBJECT);
            }
        };
        assertNotNull(new RabbitConsumer(mockRabbitConnection, EXCHANGE, SUBJECT, QUEUE, mockHandler));
    }

    @Test (expected = RabbitConnectionException.class)
    public void constructorWithError() throws IOException, RabbitConnectionException {
        new Expectations() {
            {
                mockRabbitConnection.connection(); result = mockConnection;
                mockConnection.createChannel(); result = new IOException();
            }
        };
        new RabbitConsumer(mockRabbitConnection, EXCHANGE, SUBJECT, QUEUE, mockHandler);
    }

    @Test
    public void start() throws RabbitConnectionException, IOException {
        new Expectations() {
            {
                mockRabbitConnection.connection(); result = mockConnection;
                mockConnection.createChannel(); result = mockChannel;
                mockChannel.queueDeclare(QUEUE, true, false, false, null); times = 1;
                mockChannel.queueBind(QUEUE, EXCHANGE, SUBJECT);
                mockChannel.basicConsume(QUEUE, false, withInstanceOf(DefaultConsumer.class));
            }
        };

        RabbitConsumer consumer = new RabbitConsumer(mockRabbitConnection, EXCHANGE, SUBJECT, QUEUE, mockHandler);
        consumer.start();
    }

    @Test
    public void handlerAck(@Mocked Envelope mockedEnvelope) throws IOException, RabbitConnectionException {
        new Expectations() {
            {
                mockRabbitConnection.connection(); result = mockConnection;
                mockConnection.createChannel(); result = mockChannel;
                mockChannel.queueDeclare(QUEUE, true, false, false, null); times = 1;
                mockChannel.queueBind(QUEUE, EXCHANGE, SUBJECT);
                mockHandler.handleMessage(anyString); result = MessageHandler.Response.ACK;
                mockChannel.basicAck(anyLong, false);
            }
        };

        RabbitConsumer consumer = new RabbitConsumer(mockRabbitConnection, EXCHANGE, SUBJECT, QUEUE, mockHandler);
        Deencapsulation.invoke(consumer, "handler", mockedEnvelope, message());
    }

    @Test
    public void handlerNackRequeue(@Mocked Envelope mockedEnvelope) throws IOException, RabbitConnectionException {
        new Expectations() {
            {
                mockRabbitConnection.connection(); result = mockConnection;
                mockConnection.createChannel(); result = mockChannel;
                mockChannel.queueDeclare(QUEUE, true, false, false, null); times = 1;
                mockChannel.queueBind(QUEUE, EXCHANGE, SUBJECT);
                mockHandler.handleMessage(anyString); result = MessageHandler.Response.REQUEUE;
                mockChannel.basicNack(anyLong, false, true);
            }
        };

        RabbitConsumer consumer = new RabbitConsumer(mockRabbitConnection, EXCHANGE, SUBJECT, QUEUE, mockHandler);
        Deencapsulation.invoke(consumer, "handler", mockedEnvelope, message());
    }

    @Test
    public void handlerNackReject(@Mocked Envelope mockedEnvelope) throws IOException, RabbitConnectionException {
        new Expectations() {
            {
                mockRabbitConnection.connection(); result = mockConnection;
                mockConnection.createChannel(); result = mockChannel;
                mockChannel.queueDeclare(QUEUE, true, false, false, null); times = 1;
                mockChannel.queueBind(QUEUE, EXCHANGE, SUBJECT);
                mockHandler.handleMessage(anyString); result = MessageHandler.Response.REJECT;
                mockChannel.basicNack(anyLong, false, false);
                mockRabbitConnection.publish(anyString, anyString, HandlerResponse.REJECT(), anyString);
            }
        };

        RabbitConsumer consumer = new RabbitConsumer(mockRabbitConnection, EXCHANGE, SUBJECT, QUEUE, mockHandler);
        Deencapsulation.invoke(consumer, "handler", mockedEnvelope, messageWithErrorExchange());
    }

    private byte[] messageWithErrorExchange() {
        MessageWrapper wrapper = new Message("TEST", ERROR_EXCHANGE, ERROR_SUBJECT);
        return wrapper.toJson().getBytes();
    }

    private byte[] message() {
        MessageWrapper wrapper = new Message("TEST", null, null);
        return wrapper.toJson().getBytes();
    }
}

File: src/test/scala/me/breidenbach/rabbitex/connection/RabbitConsumerTest.scala
package me.breidenbach.rabbitex.connection

import java.io.IOException

import com.rabbitmq.client.{Envelope, DefaultConsumer, Connection, Channel}
import me.breidenbach.rabbitex.MessageHandler
import me.breidenbach.rabbitex.connection.HandlerResponse.HandlerResponse
import me.breidenbach.rabbitex.test.BaseFixture

import org.mockito.{Matchers, ArgumentCaptor}
import org.mockito.Mockito._
import org.mockito.Matchers._
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.{is, nullValue, sameInstance, not}

import RabbitConsumerTest._

/**
 * Date: 10/19/14
 * Time: 2:37 PM
 * Copyright 2014 Kevin E. Breidenbach
 * @author Kevin E. Breidenbach
 */

object RabbitConsumerTest {
  val QUEUE = "queue"
  val EXCHANGE = "exchange"
  val SUBJECT = "subject"
  val ERROR_EXCHANGE = "ERROR_EXCHANGE"
  val ERROR_SUBJECT = "ERROR_SUBJECT"
  val MESSAGE_TEXT = "TEST"

  def messageWithErrorExchange(): Array[Byte] = {
    Message(MESSAGE_TEXT, Some(ERROR_EXCHANGE), Some(ERROR_SUBJECT)).toJson.getBytes
  }

  def message(): Array[Byte] = {
    Message(MESSAGE_TEXT).toJson.getBytes
  }
}

class RabbitConsumerTest extends BaseFixture {

  val mockRabbitConnection = mock[RabbitConnection]
  val mockConnection = mock[Connection]
  val mockChannel = mock[Channel]
  val mockMessageHandler = mock[MessageHandler]
  val defaultConsumerCaptor = ArgumentCaptor.forClass(classOf[DefaultConsumer])

  override def beforeEach(): Unit = {
    when(mockRabbitConnection.connection).thenReturn(mockConnection)
    when(mockConnection.createChannel()).thenReturn(mockChannel)
  }

  override def afterEach(): Unit = {
    reset(mockRabbitConnection, mockConnection, mockChannel, mockMessageHandler)
  }

  test("case class") {
    val consumer = RabbitConsumer(mockRabbitConnection, EXCHANGE, SUBJECT, QUEUE, mockMessageHandler)

    assertThat(consumer, not(nullValue()))
    verify(mockChannel, times(1)).queueDeclare(QUEUE, true, false, false, null)
    verify(mockChannel, times(1)).queueBind(QUEUE, EXCHANGE, SUBJECT)
  }

  test("case class fail") {
    when(mockConnection.createChannel()).thenThrow(new IOException("Error"))

    intercept[RabbitConnectionException] {
      RabbitConsumer(mockRabbitConnection, EXCHANGE, SUBJECT, QUEUE, mockMessageHandler)
    }
  }

  test("case class bind fail") {
    doThrow(new IOException("ERROR")).when(mockChannel).queueBind(QUEUE, EXCHANGE, SUBJECT)

    intercept[RabbitConnectionException] {
      RabbitConsumer(mockRabbitConnection, EXCHANGE, SUBJECT, QUEUE, mockMessageHandler)
    }
  }

  test("start") {
    RabbitConsumer(mockRabbitConnection, EXCHANGE, SUBJECT, QUEUE, mockMessageHandler).start()

    verify(mockChannel, times(1)).basicConsume(Matchers.eq(QUEUE), Matchers.eq(false), defaultConsumerCaptor.capture())
    assertThat(defaultConsumerCaptor.getValue.getChannel, is(sameInstance(mockChannel)))
  }

  test("handle message with ACK") {
    when(mockMessageHandler.handleMessage(MESSAGE_TEXT)).thenReturn(MessageHandler.Response.ACK)

    val mockEnvelope = mock[Envelope]
    val consumer = RabbitConsumer(mockRabbitConnection, EXCHANGE, SUBJECT, QUEUE, mockMessageHandler)
    consumer.handleMessage(mockEnvelope, message())
    verify(mockChannel, times(1)).basicAck(anyLong, Matchers.eq(false))
  }

  test("handle message with REJECT") {
    when(mockMessageHandler.handleMessage(MESSAGE_TEXT)).thenReturn(MessageHandler.Response.REJECT)

    val mockEnvelope = mock[Envelope]
    val consumer = RabbitConsumer(mockRabbitConnection, EXCHANGE, SUBJECT, QUEUE, mockMessageHandler)
    consumer.handleMessage(mockEnvelope, messageWithErrorExchange())

    verify(mockChannel, times(1)).basicNack(anyLong, Matchers.eq(false), Matchers.eq(false))
    verify(mockRabbitConnection, times(1)).publishError(ERROR_EXCHANGE, ERROR_SUBJECT, HandlerResponse.REJECT, MESSAGE_TEXT)
  }

  test("handle message with REQUEUE") {
    when(mockMessageHandler.handleMessage(MESSAGE_TEXT)).thenReturn(MessageHandler.Response.REQUEUE)

    val mockEnvelope = mock[Envelope]
    val consumer = RabbitConsumer(mockRabbitConnection, EXCHANGE, SUBJECT, QUEUE, mockMessageHandler)
    consumer.handleMessage(mockEnvelope, messageWithErrorExchange())

    verify(mockChannel, times(1)).basicNack(anyLong, Matchers.eq(false), Matchers.eq(true))
    verify(mockRabbitConnection, times(1)).publishError(ERROR_EXCHANGE, ERROR_SUBJECT, HandlerResponse.REQUEUE, MESSAGE_TEXT)
  }

  test("handle error message with REJECT and ensure no error message published") {
    when(mockMessageHandler.handleMessage("ERROR")).thenReturn(MessageHandler.Response.REQUEUE)

    val mockEnvelope = mock[Envelope]
    val consumer = RabbitConsumer(mockRabbitConnection, EXCHANGE, SUBJECT, QUEUE, mockMessageHandler)
    consumer.handleMessage(mockEnvelope, ErrorMessage("ERROR", HandlerResponse.REJECT).toJson.getBytes)

    verify(mockChannel, times(1)).basicNack(anyLong(), Matchers.eq(false), Matchers.eq(true))
    verify(mockRabbitConnection, never).publishError(anyString, anyString,
      any(classOf[HandlerResponse]), anyString)
  }

  test("default consumer") {
    when(mockMessageHandler.handleMessage(MESSAGE_TEXT)).thenReturn(MessageHandler.Response.ACK)
    val mockEnvelope = mock[Envelope]
    val consumer = RabbitConsumer(mockRabbitConnection, EXCHANGE, SUBJECT, QUEUE, mockMessageHandler)
    val defaultConsumer = consumer.defaultConsumer()

    defaultConsumer.handleDelivery("consumerTag", mockEnvelope, null, message())

    verify(mockMessageHandler, times(1)).handleMessage(MESSAGE_TEXT)
  }
}